src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -0,0 +1,678 @@
+        public static string ReplaceEncoding(string mediaType, Encoding encoding)
+        {
+            return ReplaceEncoding(new StringSegment(mediaType), encoding);
+        }
+
+        /// <summary>
+        /// Replaces the encoding of the given <paramref name="mediaType"/> with the provided
+        /// <paramref name="encoding"/>.
+        /// </summary>
+        /// <param name="mediaType">The media type whose encoding will be replaced.</param>
+        /// <param name="encoding">The encoding that will replace the encoding in the <paramref name="mediaType"/></param>
+        /// <returns>A media type with the replaced encoding.</returns>
+        public static string ReplaceEncoding(StringSegment mediaType, Encoding encoding)
+        {
+            var parsedMediaType = new MediaType(mediaType);
+            var charset = parsedMediaType.GetParameter("charset");
+
+            if (charset.HasValue &&
+                charset.Equals(encoding.WebName, StringComparison.OrdinalIgnoreCase))
+            {
+                return mediaType.Value;
+            }
+
+            if (!charset.HasValue)
+            {
+                return CreateMediaTypeWithEncoding(mediaType, encoding);
+            }
+
+            var charsetOffset = charset.Offset - mediaType.Offset;
+            var restOffset = charsetOffset + charset.Length;
+            var restLength = mediaType.Length - restOffset;
+            var finalLength = charsetOffset + encoding.WebName.Length + restLength;
+
+            var builder = new StringBuilder(mediaType.Buffer, mediaType.Offset, charsetOffset, finalLength);
+            builder.Append(encoding.WebName);
+            builder.Append(mediaType.Buffer, restOffset, restLength);
+
+            return builder.ToString();
+        }
+
+        /// <summary>
+        /// Creates an <see cref="MediaTypeSegmentWithQuality"/> containing the media type in <paramref name="mediaType"/>
+        /// and its associated quality.
+        /// </summary>
+        /// <param name="mediaType">The media type to parse.</param>
+        /// <param name="start">The position at which the parsing starts.</param>
+        /// <returns>The parsed media type with its associated quality.</returns>
+        public static MediaTypeSegmentWithQuality CreateMediaTypeSegmentWithQuality(string mediaType, int start)
+        {
+            var parsedMediaType = new MediaType(mediaType, start, length: null);
+            var enumerator = parsedMediaType._parameterEnumerator;
+
+            double quality = 1.0d;
+            while (enumerator.MoveNext())
+            {
+                if (enumerator.Current.HasName(QualityParameter))
+                {
+                    quality = double.Parse(
+                        enumerator.Current.Value.Value, NumberStyles.AllowDecimalPoint,
+                        NumberFormatInfo.InvariantInfo);
+                }
+            }
+            if (parsedMediaType.HasValue)
```

Comment:
Sorry, I was doing it wrong, i forgot I was working with a struct. I need to check enumerator.ParsingFailed not parsedMediaType.HasValue